Sources/SourceKitLSP/Swift/RewriteSourceKitPlaceholders.swift

```diff
@@ -14,34 +14,163 @@ import Foundation
 import SKLogging
 @_spi(RawSyntax) import SwiftSyntax
 
-func rewriteSourceKitPlaceholders(in string: String, clientSupportsSnippets: Bool) -> String {
-  var result = string
-  var index = 1
-  while let start = result.range(of: "<#") {
-    guard let end = result[start.upperBound...].range(of: "#>") else {
-      logger.fault("Invalid placeholder in \(string)")
-      return string
-    }
-    let rawPlaceholder = String(result[start.lowerBound..<end.upperBound])
-    guard let displayName = nameForSnippet(rawPlaceholder) else {
-      logger.fault("Failed to decode placeholder \(rawPlaceholder) in \(string)")
-      return string
+/// Translate SourceKit placeholder syntax — `<#foo#>` — in `input` to LSP
+/// placeholder syntax: `${n:foo}`.
+///
+/// If `clientSupportsSnippets` is `false`, the placeholder is rendered as an
+/// empty string, to prevent the client from inserting special placeholder
+/// characters as if they were literal text.
+func rewriteSourceKitPlaceholders(in input: String, clientSupportsSnippets: Bool) -> String {
+  var result = ""
+  var nextPlaceholderNumber = 1
+  // Current stack of nested placeholders, most nested last. Each element needs
+  // to be rendered inside the element before it.
+  var placeholders: [(number: Int, contents: String)] = []
+  let tokens = tokenize(input)
+  for token in tokens {
+    switch token {
+    case let .text(text) where placeholders.isEmpty:
+      result += text
+
+    case let .text(text):
+      placeholders.latest.contents += text
```

Comment:
Instead of having two cases that just different in a `where` clause, I would have one case with a nested `if`. I think that highlights the difference between whether we’re currently inside a placeholder better.

```suggestion
    case let .text(text):
      if placeholders.isEmpty {
        result += text
      } else {
        placeholders.latest.contents += text
      }
```